test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
+		if err != nil {
+			return err
+		}
+		newPod.Labels = newLabelMap
+		_, err = podClient.Update(newPod)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to update labels for pod %s: %v", pod.Name, err)
+	}
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// Then, verify ControllerRef for the pod is nil
+		if metav1.GetControllerOf(newPod) == nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("Failed to verify ControllerRef for the pod %s is nil: %v", pod.Name, err)
+	}
+
+	// Adoption: RS should add ControllerRef to a pod when the pod's labels change to match its labels
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// Revert to original labels so that the RS can adopt the pod again
+		newPod.Labels = labelMap
+		_, err = podClient.Update(newPod)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to revert to original labels for pod %s: %v", pod.Name, err)
+	}
+
+	controllerRef := metav1.GetControllerOf(pod)
+	if controllerRef == nil {
+		t.Fatalf("ControllerRef of pod %s is nil", pod.Name)
 	}
-	t.Logf("patched pod2 = %#v", pod2)
-	// wait for the rs to create one more pod
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 3)
+	if controllerRef.UID != rs.UID {
+		t.Fatalf("RS owner of the pod %s has a different UID: Expected %v, got %v", pod.Name, rs.UID, controllerRef.UID)
+	}
+}
+
+// Verify ControllerRef of a RS pod that has incorrect attributes is automatically patched by the RS
+func testPodControllerRefPatch(t *testing.T, c clientset.Interface, pod *v1.Pod, ownerReference *metav1.OwnerReference, rsName, ns string, expectedOwnerReferenceNum int) {
+	podClient := c.Core().Pods(ns)
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		newPod.OwnerReferences = []metav1.OwnerReference{*ownerReference}
+		_, err = podClient.Update(newPod)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to update .OwnerReferences for pod %s: %v", pod.Name, err)
 	}
-	pod2, err = podClient.Get(pod2.Name, metav1.GetOptions{})
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if metav1.GetControllerOf(newPod) != nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("Failed to verify ControllerRef for the pod %s is not nil: %v", pod.Name, err)
+	}
+
+	rs, err := c.ExtensionsV1beta1().ReplicaSets(ns).Get(rsName, metav1.GetOptions{})
```

Comment:
Had to `GET` the RS here as the function has RS's name as the parameter, not RS itself. Perhaps the function should have the RS as the parameter, not its name.

Reply:
It's better to pass in the RS so we know the UID is the same as what the caller expects, rather than something different returned by the server.

Reply:
Yes, good point. I found it hard before to convince myself to bring the whole RS (although it is just a pointer) as a function parameter - this is a good reason for similar future situations.